From aadc535e2ee5107f51273b11a2f44d8ac112d17d Mon Sep 17 00:00:00 2001
From: Mihai Moldovan <ionic@ionic.de>
Date: Tue, 29 Sep 2020 18:30:16 +0200
Subject: [PATCH] Add rpmdb in home directory to checksum and DNF context setup
 function.

The rpm package on Debian is patched to always use the RPMDB in a user's home
directory.

Tell the checksumming and DNF context setup function to prefer the rpmdb in the
user's home directory.
---
 libdnf/dnf-context.cpp | 33 +++++++++++++++++++++++++++------
 libdnf/dnf-utils.cpp   |  2 ++
 2 files changed, 29 insertions(+), 6 deletions(-)

diff --git a/libdnf/dnf-context.cpp b/libdnf/dnf-context.cpp
index d119e7de..0a343326 100644
--- a/libdnf/dnf-context.cpp
+++ b/libdnf/dnf-context.cpp
@@ -2201,12 +2201,33 @@ dnf_context_setup(DnfContext *context,
 
     /* setup a file monitor on the rpmdb, if we're operating on the native / */
     if (g_strcmp0(priv->install_root, "/") == 0) {
-        rpmdb_path = g_build_filename(priv->install_root, "var/lib/rpm/Packages", NULL);
-        file_rpmdb = g_file_new_for_path(rpmdb_path);
-        priv->monitor_rpmdb = g_file_monitor_file(file_rpmdb,
-                               G_FILE_MONITOR_NONE,
-                               NULL,
-                               error);
+        /* Try home directory first. */
+        gchar *home_dir = g_strdup(g_get_home_dir());
+
+        priv->monitor_rpmdb = NULL;
+
+        if (home_dir) {
+            rpmdb_path = g_build_filename(priv->install_root, home_dir, ".rpmdb/Packages", NULL);
+
+            g_free(home_dir);
+            home_dir = NULL;
+
+            file_rpmdb = g_file_new_for_path(rpmdb_path);
+            priv->monitor_rpmdb = g_file_monitor_file(file_rpmdb,
+                                   G_FILE_MONITOR_NONE,
+                                   NULL,
+                                   error);
+        }
+
+        if (!priv->monitor_rpmdb) {
+            rpmdb_path = g_build_filename(priv->install_root, "var/lib/rpm/Packages", NULL);
+            file_rpmdb = g_file_new_for_path(rpmdb_path);
+            priv->monitor_rpmdb = g_file_monitor_file(file_rpmdb,
+                                   G_FILE_MONITOR_NONE,
+                                   NULL,
+                                   error);
+        }
+
         if (priv->monitor_rpmdb == NULL)
             return FALSE;
         g_signal_connect(priv->monitor_rpmdb, "changed",
diff --git a/libdnf/dnf-utils.cpp b/libdnf/dnf-utils.cpp
index 874282cf..2ae179bc 100644
--- a/libdnf/dnf-utils.cpp
+++ b/libdnf/dnf-utils.cpp
@@ -32,6 +32,8 @@
 #include <errno.h>
 #include <stdlib.h>
 #include <glib/gstdio.h>
+#include <unistd.h>
+#include <pwd.h>
 
 #include "catch-error.hpp"
 #include "dnf-types.h"
-- 
2.26.2

